src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java

```diff
@@ -0,0 +1,282 @@
+        }
+        metaTileEntity.getNotifiedItemInputList().clear();
+        metaTileEntity.getNotifiedFluidInputList().clear();
+    }
+
+    @Override
+    protected void updateRecipeProgress() {
+        // todo figure out how maintenance fits in
+        int generatedSteam = this.recipeEUt * this.currentHeat / getMaximumHeat();
+        if (generatedSteam > 0) {
+            long amount = (generatedSteam + STEAM_PER_WATER) / STEAM_PER_WATER;
+            excessWater += amount * STEAM_PER_WATER - generatedSteam;
+            amount -= excessWater / STEAM_PER_WATER;
+            excessWater %= STEAM_PER_WATER;
+
+            FluidStack drainedWater = ModHandler.getWaterFromContainer(getInputTank(), (int) amount, true);
+            if (amount != 0 && (drainedWater == null || drainedWater.amount < amount)) {
+                getMetaTileEntity().explodeMultiblock();
+            } else {
+                setLastTickSteam(generatedSteam);
+                getOutputTank().fill(ModHandler.getSteam(generatedSteam), true);
+            }
+        }
+        if (currentHeat < getMaximumHeat()) {
+            setHeat(currentHeat + 1);
+        }
+
+        if (++progressTime > maxProgressTime) {
+            completeRecipe();
+        }
+    }
+
+    private int adjustEUtForConfig(int rawEUt) {
+        int throttle = ((MetaTileEntityLargeBoiler) metaTileEntity).getThrottle();
+        return Math.max(25, (int) (rawEUt * (throttle / 100.0)));
+    }
+
+    private int adjustBurnTimeForConfig(int rawBurnTime) {
+        MetaTileEntityLargeBoiler boiler = (MetaTileEntityLargeBoiler) metaTileEntity;
+        int EUt = boiler.boilerType.steamPerTick();
+        int adjustedEUt = adjustEUtForConfig(EUt);
+        int adjustedBurnTime = rawBurnTime * EUt / adjustedEUt;
+        this.excessProjectedEU += EUt * rawBurnTime - adjustedEUt * adjustedBurnTime;
+        adjustedBurnTime += this.excessProjectedEU / adjustedEUt;
+        this.excessProjectedEU %= adjustedEUt;
+        return adjustedBurnTime;
+    }
+
+    private int getMaximumHeat() {
+        return ((MetaTileEntityLargeBoiler) metaTileEntity).boilerType.getTicksToBoiling();
+    }
+
+    public int getHeatScaled() {
+        return (int) Math.round(currentHeat / (1.0 * getMaximumHeat()) * 100);
+    }
+
+    public void setHeat(int heat) {
+        if (heat != this.currentHeat && !metaTileEntity.getWorld().isRemote) {
+            writeCustomData(BOILER_HEAT, b -> b.writeVarInt(heat));
+        }
+        this.currentHeat = heat;
+    }
+
+    public int getLastTickSteam() {
+        return lastTickSteamOutput;
+    }
+
+    public void setLastTickSteam(int lastTickSteamOutput) {
+        if (lastTickSteamOutput != this.lastTickSteamOutput && !metaTileEntity.getWorld().isRemote) {
+            writeCustomData(BOILER_LAST_TICK_STEAM, b -> b.writeVarInt(lastTickSteamOutput));
+        }
+        this.lastTickSteamOutput = lastTickSteamOutput;
+    }
+
+    public void invalidate() {
```

Comment:
invalidate the current heat and the `excess...` variables as well?

Reply:
The idea behind the excess variables is to both not lose potentially gained Steam, but also to not underconsume Water and Fuel on weird burntimes, so I would rather leave these be